Block API: Add block visibility control support and UI

t-hamano opened the pull request:
First implementation for #50756

## What?

This PR adds a new block API and UI for toggling block visibility ~in the frontend~ in both the frontend and the editor..

Note: First of all, I would like to discuss why this feature is necessary and what kind of extensibility it can potentially offer in the future. Any feedback would be appreciated.

## Why?

Sometimes users may want to edit blocks in the backend without affecting the frontend, for example, a section that is not yet ready to be published.

## How?

Implementation details are as follows. Many of the specifications are similar to Block Lock.

- Add a new `blockVisibility` block support except for some generic blocks, which defaults to `true`. When a block is hidden, its state is saved in `attributes.metadata.blockVisibility`.
- Add a new `isBlockHidden` private selector to the `@wordpress/block-editor` package.
- Adds a menu item to the block settings dropdown to change the block's visibility.
- In the editor, when the `blockVisibility` attribute is `false`:
  - Visually hides the block when it or any of its child blocks are not selected.
  - Displays a button in the toolbar that can also be used to toggle visibility.
  - Displays an icon in the List View.
  - In the List view, the visibility of multiple blocks can be changed simultaneously.

## Testing Instructions

- Insert a block or pattern.
- Hide the block from the block settings dropdown.
- Deselect the block and confirm that the block is striped.
- Confirm that the block is not rendered  in the frontend

### Testing Instructions for Keyboard

In the List View, use the Tab key to focus on the hidden block. Your screen reader should announce something like this:

```
Paragraph my-anchor  link  Block 1 of 4, Level 1. Block is hidden.
```

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/14aece64-e6a7-4f22-8e60-5d0520d6991e

## Tasks

Below is a list of things I can think of that need to be addressed and discussed.

- [ ] Add core backport PR
- [ ] Add some e2e tests
- [x] Should the block be visually hidden in the editor as well as the frontend? This may cause problems with editing with the block. For example, unless the List View is open, the user will not know that there are hidden blocks there, and nothing visually should happen when, for example, moving up a block next to a hidden block via the block toolbar.
- [ ] How should we enhance this block support in the future? One of the main requests would be to be able to hide blocks based on some rulesets. You can see an example of this request in this comment: https://github.com/WordPress/gutenberg/issues/50756#issuecomment-1661142609
- [ ] If the block is hidden on the frontend, CSS and JS should not be enqueued. This may eventually be achieved in https://github.com/WordPress/wordpress-develop/pull/9213.
- [ ] List View items may already have a lot of metadata rendered. While this is not required, we may need to reorganize the information we want to display in the  List View. See https://github.com/WordPress/gutenberg/issues/59409#issuecomment-1979086745

t-hamano committed:
Block API: Add block visibility control support and UI

jasmussen commented:
Nice work. I just shared [some mockups](https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3187347960) in the main issue, which instead of dimming/overlaying the hidden block, actually hides it. In my comment I acknowledge the pros and cons to doing that, but since you already have a PR ready, I'm wondering if actually testing out the theory with something real isn't the best way to help decide. Beyond surfacing the potential need for focus handling, my hope is it'll make the behavior very obvious and intuitive in a way that opens up future opportunities. What do you think?

annezazu commented:
Working well thus far :)  Thanks for all your efforts here. It's interesting to try this with inner blocks/container blocks. I think we can learn a lot from block locking here in terms of UX:

https://github.com/user-attachments/assets/20f77692-b3df-4e2e-beda-fa9d8bfd38c1

ramonjd commented:
> which instead of dimming/overlaying the hidden block, actually hides it.

I'd suggest that both would actually be useful (also [commented](https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3172953009) something to that effect in the issue):

- Dimming/overlaying to make it obvious to the editor user which blocks are hidden and where. (or at least some non-annoying and unobtrusive way for a user to map out all hidden block locations in the block tree)
- A toggle for hiding all/individual blocks in the editor to show a preview of what the frontend _would_ look like

Future compatibility with breakpoints or other variables can be left open if we keep the data structure flexible, though @t-hamano has already [thought about this](https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3166536455). 

Oh, I don't know, something like:

```
<!-- wp:paragraph {
	"metadata":{
		"blockVisibility": [
            {
				"type": "breakpoint",
				"value": "var:preset|breakpoint|small",
			}
          ],
	}
} -->
...
```

t-hamano committed:
Merge branch 'trunk' into block-visibility

t-hamano committed:
Render hidden block block

t-hamano committed:
Remove unused code

t-hamano commented:
I've been experimenting with visually hiding blocks in the editor.

My first attempt was to simply not render the block when it's not selected. As expected, this caused some issues:

- There was no visual change when moving a block up or down.
- The insertion indicator didn't work properly when dragging a block, and I couldn't move the block.

<details><summary>Video</summary>

https://github.com/user-attachments/assets/74be7986-663e-4270-b796-042488355227

</details> 


To solve this, I tried the following approach:

- Hide the block when it's not selected.
- However, render a small placeholder when a sibling block is selected or while dragging any block.

<img width="550" height="149" alt="image" src="https://github.com/user-attachments/assets/db709a22-5834-4eeb-bbcf-5c346bf774da" />

This should solve the two problems mentioned above.

The technical details of the approach are as follows:

- Added a [`isSiblingBlockSelected` private selector](https://github.com/WordPress/gutenberg/pull/71203/files#diff-23e851c3077ed95d23bb1669da8bafbae3f1e2e952897987e45b49555004dd83R681) to check if a sibling block is selected.
- Added a new [internal `HiddenBlockPlaceholder` component](https://github.com/WordPress/gutenberg/pull/71203/files#diff-1a2c50ed3ed41e109ae126fe42002702f6866379ed7770890a5d0efa8b36d320R29) for placeholders.

Let me know what you think!

https://github.com/user-attachments/assets/9a078e43-494c-4994-90cb-2d3bc7ae1b8a

t-hamano committed:
Always hide block when it's not selected

t-hamano committed:
Extend existing isBlockVisible selector and use it

t-hamano committed:
Revert "Extend existing isBlockVisible selector and use it"

This reverts commit c3f71355a4562be2847340d0e05c06f3a36c33c3.

t-hamano committed:
Remove SCSS import statement

t-hamano committed:
Use "isBlockHidden"

t-hamano committed:
Remove useBlockVisibility hook

t-hamano committed:
useBlockDropZone: Filter out blocks that are hidden

t-hamano committed:
Remove useBlockVisibility hook

t-hamano committed:
Update doc

t-hamano committed:
Show hidden block placeholder when dragging

t-hamano commented:
> Although I can be tempted to start with less than this, the behavior here feels pretty reasonable.

I like to start with a minimal implementation and iterate, so I don't mind removing this placeholder 👍

> Maybe we should deprecate it for a private, renamed alternative (`isBlockInViewport`).

I agree. Based on [the search results on WPdirectory](https://wpdirectory.net/search/01K2ZV00A4MF0ABAQMTXMYJN9C), it seems like almost no consumers are using this API.

t-hamano committed:
Revert "Show hidden block placeholder when dragging"

This reverts commit 2bfa54e618a6f6f3df1646cf31bf7b9d3493ea3f.

t-hamano committed:
Show block toolbar button only when editing mode is default

t-hamano commented:
> I noticed one thing when editing a page/post's template. You can hide template blocks such as post title, and save while editing the template.
> 
> When previewing the template, I can still toggle show/hide, but it doesn't create a dirty state and therefore can't be saved. To reproduce, first hide a block in template editing mode, save, then switch back to the post or page:

Good catch! 8c5c6257f4071827f05e36e2f897005ac3693949 should have fixed the problem.

supernovia commented:
> You can test it with the latest commit

This is helpful! Should there be some kind of indicator when using the arrow buttons to move blocks, too? (note I didn't test the newer version yet)

https://github.com/user-attachments/assets/8de2e6b9-8114-4e11-aad2-f28b3c0d4cd3

tyxla reviewed (commented):
Thanks for working on this one @t-hamano!

I'm yet to test/review this a bit more thoroughly, but came to say a few things.

I think the toolbar hide/show button appears a bit inconsistently right now:


https://github.com/user-attachments/assets/84597f34-9d86-451f-b816-c07b3546d38d

Should the hidden blocks have an impact on the post outline:

<img width="185" height="95" alt="Screenshot 2025-08-19 at 20 23 30" src="https://github.com/user-attachments/assets/78a4ec2e-8254-43f1-a449-a0d1993b71e9" />

I also think we might need an additional indication on the block itself that it's hidden when currently selected. I don't think the toolbar button is enough:

<img width="457" height="124" alt="Screenshot 2025-08-19 at 20 25 52" src="https://github.com/user-attachments/assets/0f2ff403-76e7-465e-a15a-8a8b714a548c" />

t-hamano commented:
@tyxla Thanks for the review!

> I think the toolbar hide/show button appears a bit inconsistently right now:

This is the intentional behaviour: This is to prevent focus loss, and is the same as [the implementation of the block lock button](https://github.com/WordPress/gutenberg/blob/a1a76239accd46c54952895adbddbc2878a409c9/packages/block-editor/src/components/block-lock/toolbar.js#L25-L34).

> Should the hidden blocks have an impact on the post outline:

I'd like to try whether we can omit the hidden contents when calculating characters, words, and time to read.

> I also think we might need an additional indication on the block itself that it's hidden when currently selected. I don't think the toolbar button is enough:

I don't have any ideas yet for distinguishing hidden blocks, but one idea might be to change the block icon color, like the connected block:

<img width="298" height="118" alt="image" src="https://github.com/user-attachments/assets/95779c94-0eb0-436c-a149-a9e0761e692e" />

Mamaduka commented on packages/block-editor/src/components/block-visibility/menu-item.js:
```diff
@@ -0,0 +1,47 @@
+			return select( blockEditorStore ).getBlocksByClientId( clientIds );
+		},
+		[ clientIds ]
+	);
+
+	const hasHiddenBlock = blocks.some(
+		( block ) => block.attributes.metadata?.blockVisibility === false
+	);
+
+	const toggleBlockVisibility = () => {
+		blocks.forEach( ( block ) => {
+			updateBlockAttributes( block.clientId, {
+				metadata: cleanEmptyObject( {
+					...block.attributes?.metadata,
+					blockVisibility: hasHiddenBlock ? undefined : false,
+				} ),
+			} );
+		} );
+	};
```
The `updateBlockAttributes` can process multiple blocks without a loop. The similar non-batched actions have their problems; they could be slow and create multiple history entries (undos).

t-hamano replied:
Fixed: fb2d6688942714eacc4bb9eed6685a00b2ff6ce9

This process keeps other `attributes.metadata` properties, like `name` property:

https://github.com/user-attachments/assets/f48b7397-4456-42c3-bc1f-a59507ca4ed0

Mamaduka replied:
Thanks, @t-hamano!

Mamaduka commented on docs/reference-guides/block-api/block-supports.md:
````diff
@@ -174,6 +174,32 @@ attributes: {
 }
 ```
 
+## blockVisibility
````
Maybe we should change this to `visibility` and the block part, which seems redundant.

P.S. The block editor already uses similar terminology, `isBlockVisible` and `setBlockVisibility`, to track if the block is visible in the canvas viewport. It would be nice to avoid confusion, but I can't think of better terminology.

ramonjd replied:
🤔 I forgot about `isBlockVisible` - thanks for pointing it out. I [originally suggested something other than visibility](https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3177398128) as it's too close the name of a CSS property, and could clash semantically if it block supports ever gets it.

We could remove the ambiguity and select a different word. The current `isBlockVisible` tells the system whether the block is visible, whereas we're talking about a user-driven setting. What does the user want to do? _temporarily_ hide blocks (otherwise they'd be deleted) and to stop them rendering on the frontend. It's a bit nuanced, swinging between _ prevent from being seen_ and _switch off_. 

Some alternatives

- blockDisable/disableBlock
- blockMasking
- blockConceal
- blockObscurity
- blockCloaking
- blockSuppress

There might be a PhD thesis here somewhere 😆

t-hamano replied:
I don't have a strong opinion on which terminology (`visibility` or `blockVisibility`) is better, but the existing `isBlockVisible` is indeed confusing. There is a proposal to rename the API: https://github.com/WordPress/gutenberg/pull/71203#issuecomment-3195081517

tyxla commented:
> > I think the toolbar hide/show button appears a bit inconsistently right now:
> 
> This is the intentional behaviour: This is to prevent focus loss, and is the same as [the implementation of the block lock button](https://github.com/WordPress/gutenberg/blob/a1a76239accd46c54952895adbddbc2878a409c9/packages/block-editor/src/components/block-lock/toolbar.js#L25-L34).

Hmm, I see, but the disappearing block toolbar button seems broken to me. If we choose to render a button in the block toolbar, I think it should appear there consistently, just like it does in the block settings dropdown.

> 
> > I also think we might need an additional indication on the block itself that it's hidden when currently selected. I don't think the toolbar button is enough:
> 
> I don't have any ideas yet for distinguishing hidden blocks, but one idea might be to change the block icon color, like the connected block:

Perhaps we could play with the opacity of the rendered block, just like we do for the placeholder title/block:

|Code|Visual|
|---|---|
|<img width="317" height="61" alt="Screenshot 2025-08-20 at 14 22 54" src="https://github.com/user-attachments/assets/4a4b3901-30f9-48aa-9979-f5e4e5217843" />|<img width="308" height="164" alt="Screenshot 2025-08-20 at 14 22 57" src="https://github.com/user-attachments/assets/fe759d51-ef7b-4054-b81f-19f596c7e94f" />|

Or we could display a "crossed-out eye" icon somewhere in one of the corners.

I wonder if @WordPress/gutenberg-design folks have suggestions here.

annezazu commented:
There's something odd happening when duplicating blocks, which I'd expect folks to do especially when trying out different designs. Basically, when I duplicate a block that I've hidden, it doesn't appear in the editor unless I change it to "show". I'd expect I'd still be able to see it:

https://github.com/user-attachments/assets/c4345b28-6b29-4812-8175-54e9202659cc

t-hamano committed:
Merge branch 'trunk' into block-visibility

t-hamano committed:
Process metadata updates for multiple blocks at once

t-hamano commented:
> There's something odd happening when duplicating blocks, which I'd expect folks to do especially when trying out different designs. Basically, when I duplicate a block that I've hidden, it doesn't appear in the editor unless I change it to "show".

Technically, this is correct, since blocks are duplicated with all of their content, styles, and metadata.

Personally, the current behavior feels natural to me. For example, if you copy a hidden layer in Figma, the copied layer remains hidden.

https://github.com/user-attachments/assets/4edffcb6-30da-4cfb-b616-efff093fe187

t-hamano committed:
Preserve block visibility metadata in block transforms

t-hamano commented:
> if the hidden/lock indicators in the block toolbars have to be present always, we should remove them entirely.

If I had to choose, I too would prefer to remove the button from the toolbar entirely.

tyxla commented:
> > if the hidden/lock indicators in the block toolbars have to be present always, we should remove them entirely.
> 
> If I had to choose, I too would prefer to remove the button from the toolbar entirely.

Same here. 

We could also introduce a keyboard shortcut for toggling the visibility.

t-hamano committed:
Merge branch 'trunk' into block-visibility

t-hamano committed:
Remove support from schema and docs

t-hamano committed:
Add experiments setting

t-hamano committed:
Make Block Visibility as experiment feature

t-hamano commented:
Sorry for the late reply. 

I've decided to mark this feature as experimental so that we can continue to improve it and determine whether it's ready to ship in 6.9.

<img width="572" height="142" alt="image" src="https://github.com/user-attachments/assets/34fb2523-d456-4a0a-ac89-3c0fa2d53e73" />

The support name has the `__experimental` prefix because we won't be exposing this support in [the core block reference](https://github.com/WordPress/gutenberg/blob/trunk/docs/reference-guides/core-blocks.md). Once we stabilize this new API, we plan to remove the `__experimental` prefix.

I'd be grateful for any feedback on whether this feature can be shipped as an experimental setting, whether there's something wrong with the fundamental approach, and whether there are any issues that need to be resolved before shipping.

t-hamano committed:
Rename __experimentalBlockVisibility to blockVisibility

ramonjd commented on packages/block-editor/src/components/block-visibility/menu-item.js:
```diff
@@ -0,0 +1,53 @@
+			blocks?.map( ( { clientId, attributes } ) => [
+				clientId,
+				{
+					metadata: cleanEmptyObject( {
+						...attributes?.metadata,
+						blockVisibility: hasHiddenBlock ? undefined : false,
+					} ),
```
I'm not sure if it's here but it looks like unsynced patterns are "detached" when toggling visibility:


https://github.com/user-attachments/assets/49db272e-5eef-4a35-aef2-6ac4f9a290ca

ramonjd replied:
Oh, to clarify, it only happens for me when toggling from the toolbar.

talldan replied:
Yeah, can confirm - it looks like the other `metadata` attributes get stripped when toggling via the toolbar.

t-hamano replied:
Fixed in 3b152ca06bbbd6915f1eb88c1d09aa6a1d7b2cf9

ramonjd reviewed (commented):
I retested and things are working well for me still. The list view + toolbar icon is, I believe, satisfactory as a minimum starting point.

Some things I noted just to make sure they're expected:

- multiselection doesn't show visibility icon in the toolbar or context menu when selecting two invisible blocks
- I just had a question about whether patterns and toolbar toggling should detach the pattern (I don't think so?)

I also tested with the [patterns + content only experiment](https://github.com/WordPress/gutenberg/issues/71517) on and couldn't see any major conflicts.

I will gladly defer to @jasmussen on the finer points here.

> I've decided to mark this feature as experimental so that we can continue to improve it and determine whether it's ready to ship in 6.9.

Good option to reduce the risk. Given it's behind an experimental flag I think it's valid to merge and iterate. 

> How should we enhance this block support in the future?

Adding some, perhaps whacky, ideas for future follow ups:

- Visual indication on the canvas that a block/pattern is hidden
- Bulk operations like select all hidden blocks, show/hide by block type
- Add performance tests (100+ blocks)
- Not sure if this would be useful, but an API for third party devs to control conditional visibility support, e.g., based on rules like user roles, device types, or block types 
- Will these need  any further theme.json support?

Once again, thanks @t-hamano for all the effort.

talldan commented on packages/block-editor/src/components/block-visibility/toolbar.js:
```diff
@@ -0,0 +1,76 @@
+								blockVisibility: newBlockHidden
+									? false
+									: undefined,
+							},
```
This looks like where the bug Ramon spotted happens. Need to spread in any existing `metadata` props as part of the attributes update.

ramonjd replied:
Good spotting!

t-hamano replied:
Fixed in https://github.com/WordPress/gutenberg/commit/3b152ca06bbbd6915f1eb88c1d09aa6a1d7b2cf9

sunyatasattva commented on packages/block-library/src/block/block.json:
```diff
@@ -24,6 +24,7 @@
 		"html": false,
 		"inserter": false,
 		"renaming": false,
+		"__experblockVisibilityimentalBlockVisibility": false,
```
Something seems to have happened to this line of code 😬

t-hamano replied:
Fixed in 1c9b4074081d428984cc24c4208834a51d66292f

sunyatasattva commented:
Thanks for working on this, this is a great step! I've tried a bunch of strange edge cases and found some potential issues, also I've got some questions and comments.

## Issues

### Strange behavior with the Query Loop

https://github.com/user-attachments/assets/4e44d5db-7d5b-4c43-802d-4e6ade71edae

If you try to hide a block inside the Query Loop Post Template, the behavior of hiding is erratic. The expectation is that all blocks of that type will be hidden, but only the currently selected block gets hidden in the editor.

### Interaction with the Accordion block

The [new experimental Accordion block has been recently added](https://github.com/WordPress/gutenberg/pull/64119). This might be a weird interaction or edge case that we might not want to support anyways, but if I hide the title of an accordion item, the content will always be displayed (while the expectation is that content is collapsed under the item).

<img width="1162" height="284" alt="Screenshot 2025-09-25 at 17 25 31" src="https://github.com/user-attachments/assets/c4b5fea6-e872-4e3b-a0d0-5d2535c50bc8" />

I wonder if the block visibility option should interact in any way with the locking of blocks (i.e. blocks need to be unlocked before the visibility can be changed).

## Generic feedback

1. In the list view, when the icon is added next to the three dots and there is a separator, the icons look a bit visually unbalanced, see below.

<img width="65" height="43" alt="Screenshot 2025-09-25 at 17 15 37" src="https://github.com/user-attachments/assets/ce461c82-183a-4ef4-b778-59068206258b" />

2. Also, in the same view, I'd expect the eye icon to be clickable and visibility to be toggleable from there. I understand this might be a personal expectation, and the lock icon is not clickable, but in all design software I have ever used, visibility is toggleable from the layers panel through the click of this icon, so I find it weird.

3. I agree with the commenters above that the expectation is that something would _immediately_ happen when a block is hidden. Likely, the block should immediately disappear. Alternatively, the opacity can be reduced. But right now there is no indication of anything happening until I click away after I set the visibility, which breaks user expectations.

## Questions

Are we opening the door to further and more complex use-cases with this API? Obviously, the most requested upgrade to this would be conditional rendering. While this PR covers one interesting use-case (which is experimenting with design), the potential is to be more than just a design tool, and become an actual editorial tool. The upgrade could be iterative, but the `blockVisibility` attribute could eventually accept an object, instead of just a boolean, that defines the conditions of the visibility.

That would be helpful for many use-cases, such as viewport-based or user-permission-based visibility.

t-hamano committed:
Merge branch 'trunk' into block-visibility

t-hamano committed:
Fix: other metadata attributes get stripped when toggling via the toolbar

t-hamano committed:
Show block toolbar button when multiple blocks are selected

t-hamano committed:
Fix incorrect block.json definition

t-hamano commented:
Thanks everyone for the reviews!


> multiselection doesn't show visibility icon in the toolbar or context menu when selecting two invisible blocks

Fixed in fffcf35225

> I just had a question about whether patterns and toolbar toggling should detach the pattern (I don't think so?)

I think the problem has been fixed in 3b152ca06bbbd6915f1eb88c1d09aa6a1d7b2cf9 (Toggling the visibility button on the toolbar does not detach the pattern).

> - Visual indication on the canvas that a block/pattern is hidden
> - Bulk operations like select all hidden blocks, show/hide by block type
> - Add performance tests (100+ blocks)

Yes, this could be considered as a follw-up.

> - Not sure if this would be useful, but an API for third party devs to control conditional visibility support, e.g., based on rules like user roles, device types, or block types 

Since the block visibility feature is just block support and attributes, I think consumers can add their own rules with appropriate filters either from the server side or the client side.

> - Will these need  any further theme.json support?

 I'm not considering this at the moment, but if there is a strong need for it, we may consider it in the future.

t-hamano committed:
Apply opacity to hidden and selected blocks

ramonjd commented:
> multiselection doesn't show visibility icon in the toolbar or context menu when selecting two invisible blocks
> I just had a question about whether patterns and toolbar toggling should detach the pattern (I don't think so?)

Just checked these again. LGTM Great work!

> If you try to hide a block inside the Query Loop Post Template, the behavior of hiding is erratic. The expectation is that all blocks of that type will be hidden, but only the currently selected block gets hidden in the editor.

I can't seem to reproduce this one, maybe the latest commits fixed it? 🤔

ramonjd reviewed (approved):
LGTM

I know there's a [pending opacity change](https://github.com/WordPress/gutenberg/pull/71203#issuecomment-3350262274), but I'll tentatively approve since it's behind an experiment. 

This PR is large so it's a solid foundation on which to iterate.

Thanks for all the work @t-hamano

t-hamano committed:
Revert "Apply opacity to hidden and selected blocks"

This reverts commit 5596f831580a507fac32408fce5a9c3507a056cc.

t-hamano committed:
Provide blockVilibility metadata to preview context

t-hamano committed:
Accordion Header, Accordion Panel: disable blockVisibility support

jasmussen commented:
> I'd like to understand this in more detail. So does this mean that we need to implement the following in this PR?

For this PR, I think we either need the opacity change or the change Jay is proposing. And which one depends on the technical questions, so I defer to you on that. So to be more precise, how might we accomplish what he's proposing? Here's what my CSS awareness can come up with:

* 0px height, so the block is still there and the block toolbar is still there. Might cause a layout shift still when you deselect.
* `display: none;`—but I'm not sure if that works or not. If you truly hid the block, what would happen to focus, and to the block toolbar?

And are there any other options? The ideal really is: you hide the block, and it's gone, but you don't lose focus and the block toolbar is still there until you select another block. 

What do you think?

t-hamano committed:
Visually hide block

talldan commented:
> Personally, I'm not yet 100% confident about shipping this feature in 6.9 😅 Once GB 21.8 is released, I'll try to solicit feedback in our outreach channels to see if there are any major issues. After that, I want to consider stabilizing it in GB 21.9. What do you think?

I think it's more likely to get more widespread testing and quicker feedback if there's no experiment flag, so if it were me I'd probably aim for removing the experiment.

I personally think of experiments as things where there are known issues or experience gaps, but they largely work. This feature feels relatively complete and ready for wider testing.

t-hamano committed:
Merge branch 'trunk' into block-visibility

t-hamano committed:
Stabilize feature

t-hamano committed:
Add backport changelog

t-hamano committed:
Update server side function

t-hamano committed:
Use "visibility:hidden" to visually hide blocks

t-hamano commented:
Ok, let's stabilize the feature with this PR. I made the following changes:

- bb9ec3702fd1f6b385c1a3f47001b9a3f5c0c014: Stabilize feature
- 4c7338c254e7ff00325c1e6cce7577ccd3195ae5: Submit a core backport PR and add a backport changelog
- 861689cb3c2e14f43da70f80a7efbc1c17b28eba: Add missing block support checks on the server side
- 09295598018123e0ac6b5e328c3fd3412827bd9b: Use `visibility:hidden` to visually hide blocks. I believe this is a more fundamental approach and the block toolbar seems to work fine.

talldan commented on packages/block-editor/src/components/block-visibility/toolbar.js:
```diff
@@ -0,0 +1,91 @@
+import { cleanEmptyObject } from '../../hooks/utils';
+
+export default function BlockVisibilityToolbar( { clientIds } ) {
+	const blocks = useSelect(
+		( select ) => {
+			return select( blockEditorStore ).getBlocksByClientId( clientIds );
+		},
+		[ clientIds ]
+	);
+	const canToggleBlockVisibility = useSelect(
+		( select ) => {
+			const { getBlockName, getBlocksByClientId } =
+				select( blockEditorStore );
+			const _blocks = getBlocksByClientId( clientIds );
+			return _blocks.every( ( { clientId } ) =>
+				hasBlockSupport(
+					getBlockName( clientId ),
+					'blockVisibility',
+					true
+				)
+			);
+		},
+		[ clientIds ]
+	);
```
The `useSelect` calls are both pretty similar, they could be combined into one to reduce subscriptions, or you could use `blocks` from the first `useSelect` in the second.

t-hamano replied:
Good catch, fixed in dd0d1cb4dfaf80a38c9db55c9a3f6f1fb9774460

talldan commented on packages/block-library/src/block/block.json:
```diff
@@ -24,6 +24,7 @@
 		"html": false,
 		"inserter": false,
 		"renaming": false,
+		"blockVisibility": false,
```
I'm curious about the reasoning for not allowing synced patterns to be hidden. I think that might be quite a handy use case (same for template parts which are essentially another form of pattern).

It's something that can be revised later quite easily so doesn't necessarily need to be addressed in this PR.

t-hamano replied:
There's no clear reason for this at the moment 😅 I've disabled this support for generic blocks just to be safe, but I'll consider addressing it in a follow-up.

talldan reviewed (approved):
LGTM, and it's working nicely. I left a couple of small comments, but the code is really tidy. E2E tests would be nice, but can be a separate PR.

One thing I notice when testing is that some transforms might not retain the block visibility, others do. Paragraph to Preformatted seems to make the block visible again, but Preformatted to Paragraph keeps the block hidden. Not sure why, but seems like a real edge case, so again not urgent, but maybe something to fix at some point.

jasmussen commented:
This is the current state:

![state](https://github.com/user-attachments/assets/c696af65-9e46-4834-af8f-0b3418e3d19c)

I have to admit, I preferrred the 0 height one quite a bit. Could we have made that also 0 width to address Dan's good note?

Not suggesting this as a blocker. I think the PR should land. But I'd like, if possible, to not entirely close the door on the collapsing, because for the moment, it's still collapsing when you deselect the block, so it's arguably more confusing now than when at least the block appeared to be removed from the flow (which is the case on the frontend).

t-hamano commented:
> One thing I notice when testing is that some transforms might not retain the block visibility, others do

Ah, I think this problem is related to this issue:

- https://github.com/WordPress/gutenberg/issues/71922

This issue will likely occur with any blocks that extract only specific attributes and call createBlock within the transform property's callback function.

We may need to fix this in a follow-up so that all three metadata attributes (`name`. `blockVisibility`, and `commentId`) are not lost.

t-hamano committed:
Combine multiple useSelect call into one

t-hamano committed:
Reset styles from hidden blocks

t-hamano commented:
> I have to admit, I preferred the 0 height one quite a bit.

I agree. Fixed in 96428fdea27819828af1ec1b9bfeb9db26a7e29f

> Could we have made that also 0 width to address Dan's good note?

I tried it, but since blocks in the editor are horizontally centered, the toolbar movement can be annoying:

![08a36e833a0a0fe8b8341eb03be70b27](https://github.com/user-attachments/assets/a784c8c6-7a94-430d-bb3f-5e0645007ae8)

I'll try to find a better approach to hiding selected blocks in a follow-up.

t-hamano commented:
> Another issue I spotted with both `width: 0` and `height: 0` is that the sibling block inserter tries to match the width of the hidden block in a vertical layout, and it looks noticeably strange. Maybe there are ways to solve that (e.g. by not trying to reference the hidden block's width, but the nearest visible block instead).

Yes, I'm investigating this...

mukeshpanchal27 commented on lib/block-supports/block-visibility.php:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Render nothing if the block is hidden.
+ *
+ * @param  string $block_content Rendered block content.
+ * @param  array  $block         Block object.
+ * @return string                Filtered block content.
```
```suggestion
 * @param string $block_content Rendered block content.
 * @param array  $block         Block object.
 * @return string Filtered block content.
```

Per WPCS https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods

t-hamano committed:
Update lib/block-supports/block-visibility.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

t-hamano commented:
I understand what the ideal behavior here would be, but I haven't yet figured out how to technically achieve all of the following:

- We can't apply `display:none` in order to render the block toolbar. We need to render some element.
- For vertical layout, the height should be set to 0 and the width should match the adjacent block.
- For horizontal layout, the width should be set to 0 and the height should match the adjacent block.
- Make sure the hoge inserter is displayed properly

jasmussen commented:
I want to briefly pause and thank you for your work. It's been inspiring, it's been great. 

Then I want to echo that this is solid work, I think it should land. To me it even feels 6.9 ready. Now whether we implement the visibility or the opacity change as you had earlier and land this, then explore height/width/collapsing separately, I'll defer to you all. To be very clear, I think it's an important value add, worth pursuing. But I also don't think it is _required_ for the first iteration to land, especially as the block really does truly disappear when you deselect the block. That feels the main constraint.

t-hamano committed:
Better behavior for hidden blocks

jasmussen commented:
(⊃｡•́‿•̀｡)⊃━☆.*･ magic

tyxla commented:
Thanks for the great work on this one @t-hamano 🫶

enodekciw commented:
Block visibility manager of some sort should be in Gutenberg, yes. However, current implementation looks questionable, imo. 

- as per Why? - "Sometimes users may want to edit blocks in the backend without affecting the frontend, for example, a section that is not yet ready to be published." - how users are supposed to do that, if block gets hidden in the Editor? Unhide it, edit it, then if it's still not final, hide it again and save the post? Basically, keep hiding and unhiding it while doing the edits?
- How useful is hide/show option with no conditional settings? Basically, it only allows users to "draft" blocks without affecting the frontend and not much else. I've been using Gutenberg in production since 2019 and I could count on my fingers the times I needed plain hide/show block functionality with no conditional settings. Usually, when I need to manage block visibility, I need to tie it to some sort of conditional state like user role, query string, screen size, or so.
- Now it's saved into `attributes.metadata.blockVisibility`, are there any other visibilities planned to be saved in metadata except block visibility? Isn't `block` in attribute name a bit redundant? I mean, block label is saved into `attributes.metadata.name` not `attributes.metadata.blockName`, so why not just use `attributes.metadata.visibility` for visibility? :) 

I'm extensively using https://github.com/ndiego/block-visibility plugin for managing block visibility state. Plugin seems to be really solid, open-source, somewhat easy to extend. It also allows not only basic show/hide but also a lot of other different options to manage the visibility (by query strings, screen sizes, user roles, etc.). I wished core Gutenberg implementation could borrow some ideas from a plugin like this :)

t-hamano commented:
@burnuser @enodekciw Thanks for the feedback.

> Follow your own target declaration and make the block invisible ONLY in the frontend.
> If you want some extra, mark such a block in the editor, e.g. with a special wireframe or an icon, which declares its invisible frontend status at a glance.

> how users are supposed to do that, if block gets hidden in the Editor? Unhide it, edit it, then if it's still not final, hide it again and save the post? Basically, keep hiding and unhiding it while doing the edits?

There is a discussion about this here: https://github.com/WordPress/gutenberg/pull/71203#issuecomment-3350354993

In many design tools, hiding a layer actually makes that layer disappear. If you want to edit a layer, you need to explicitly show that layer. Also, in order to see how it will ultimately be displayed on the front end, the block needs to be hidden in the editor as well. That's my understanding, but I agree that there are certainly many steps involved in editing the content of a hidden block.

> How useful is hide/show option with no conditional settings? Basically, it only allows users to "draft" blocks without affecting the frontend and not much else. I've been using Gutenberg in production since 2019 and I could count on my fingers the times I needed plain hide/show block functionality with no conditional settings. Usually, when I need to manage block visibility, I need to tie it to some sort of conditional state like user role, query string, screen size, or so..

I think this needs to be carefully considered based on various user feedback. This is because it is difficult to remove a feature once it has been added. We have a responsibility to ensure full backward compatibility.

> Now it's saved into `attributes.metadata.blockVisibility`, are there any other visibilities planned to be saved in metadata except block visibility? Isn't `block` in attribute name a bit redundant? I mean, block label is saved into `attributes.metadata.name` not `attributes.metadata.blockName`, so why not just use `attributes.metadata.visibility` for visibility? :) 

One reason is to avoid confusion with the CSS property "visibility". See https://github.com/WordPress/gutenberg/issues/50756#issuecomment-3177398128

> I'm extensively using https://github.com/ndiego/block-visibility plugin for managing block visibility state. Plugin seems to be really solid, open-source, somewhat easy to extend. It also allows not only basic show/hide but also a lot of other different options to manage the visibility (by query strings, screen sizes, user roles, etc.). I wished core Gutenberg implementation could borrow some ideas from a plugin like this :)

I'm also aware of that plugin. We can consider which ideas to incorporate as part of future enhancements.

If you have any specific ideas about what kind of feature enhancements are needed, we would welcome you to submit a new issue.
